openmls/src/binary_tree/array_representation/diff.rs

```diff
@@ -171,66 +161,43 @@ impl<'a, T: Clone + Debug> AbDiff<'a, T> {
+        let b_iter = Box::new(original_leaves) as Box<dyn Iterator<Item = (LeafNodeIndex, &T)>>;
 
         // We only compare indices, not the actual leaves
-        let cmp = |&(x, _): &(u32, &T)| x;
+        let cmp = |&(x, _): &(LeafNodeIndex, &T)| x;
 
         sorted_iter(a_iter, b_iter, cmp, self.size as usize)
     }
 
     // Functions related to the direct paths of leaves
     //////////////////////////////////////////////////
 
-    /// Returns a vector of [`NodeIndex`] instances, each one referencing a
-    /// node in the direct path of the given [`LeafIndex`], ordered from the
+    /// Returns a vector of [`ParentNodeIndex`] instances, each one referencing a
+    /// node in the direct path of the given [`LeafNodeIndex`], ordered from the
     /// parent of the corresponding leaf to the root of the tree.
-    pub(crate) fn direct_path(
-        &self,
-        leaf_index: LeafIndex,
-    ) -> Result<Vec<NodeIndex>, OutOfBoundsError> {
-        let node_index = to_node_index(leaf_index);
-        // `direct_path` only throws an error if the input index is out of bounds.
-        direct_path(node_index, self.tree_size()).map_err(|_| OutOfBoundsError::IndexOutOfBounds)
+    pub(crate) fn direct_path(&self, leaf_index: LeafNodeIndex) -> Vec<ParentNodeIndex> {
+        direct_path(leaf_index, self.size())
     }
 
     /// Sets all nodes in the direct path to a copy of the given node. This
     /// function will throw an [`ABinaryTreeDiffError::OutOfBounds`] error if
     /// the given index does not correspond to a leaf in the diff.
-    pub(crate) fn set_direct_path_to_node(
-        &mut self,
-        leaf_index: LeafIndex,
-        node: &T,
-    ) -> Result<(), ABinaryTreeDiffError> {
-        let node_index = to_node_index(leaf_index);
-        let direct_path = direct_path(node_index, self.tree_size())?;
+    pub(crate) fn set_direct_path_to_node(&mut self, leaf_index: LeafNodeIndex, node: &T) {
+        let direct_path = self.direct_path(leaf_index);
         for node_index in &direct_path {
-            self.replace_node(*node_index, node.clone())?;
+            self.replace_node(TreeNodeIndex::Parent(*node_index), node.clone());
         }
-        Ok(())
     }
 
     /// Sets the nodes in the direct path of the given leaf index to the nodes
     /// given in the `path`.
-    ///
-    /// Returns an error if the given `leaf_index` does not correspond to a leaf
-    /// in the diff or if the given `path` does not have the same length as the
-    /// leaf's direct path.
-    pub(crate) fn set_direct_path(
-        &mut self,
-        leaf_index: LeafIndex,
-        path: Vec<T>,
-    ) -> Result<(), ABinaryTreeDiffError> {
-        let node_index = to_node_index(leaf_index);
-        let direct_path = direct_path(node_index, self.tree_size())?;
-        if path.len() != direct_path.len() {
-            return Err(ABinaryTreeDiffError::PathLengthMismatch);
-        }
```

Comment:
This check goes away, but this function is crate-public and it's not clear where this check is made instead. If we want to remove this Result, we could have a `Path` type that gives us guarantees w.r.t. its length.

Reply:
That should be covered by `ValSem202`.